File: src/javascript/datastructures/queue/2.queue_with_linkedlist/queue_with_linkedlist.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class MyQueue {
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }

    // Core queue operations
    enqueue(value) {
        const newNode = new Node(value);
        if (this.isEmpty()) {
            this.first = newNode;
            this.last = newNode;
        } else {
            this.last.next = newNode;
            this.last = newNode;
        }
        this.length++;
    }

    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Queue Underflow: Cannot dequeue from an empty queue.");
        }

        const temp = this.first;
        if (this.length === 1) {
            this.first = null;
            this.last = null;
        } else {
            this.first = this.first.next;
            temp.next = null;
        }

        this.length--;
        return temp.value;
    }

    peek() {
        if (this.isEmpty()) {
            throw new Error("Queue Underflow: Cannot peek from an empty queue.");
        }
        return this.first.value;
    }

    // Helper functions
    isEmpty() {
        return this.length === 0;
    }

    size() {
        return this.length;
    }

    printQueue() {
        let current = this.first;
        const values = [];
        while (current) {
            values.push(current.value);
            current = current.next;
        }
        return values.join(' ');
    }
}

// Usage example
function useQueue() {
    const queue = new MyQueue();

    console.log("Is empty?", queue.isEmpty()); // true

    try {
        queue.dequeue(); // Should throw
    } catch (e) {
        console.error(e.message);
    }

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    console.log("Queue contents:", queue.printQueue()); // 1 2 3
    console.log("Front element:", queue.peek()); // 1
    console.log("Dequeued element:", queue.dequeue()); // 1
    console.log("Queue after dequeue:", queue.printQueue()); // 2 3
}

useQueue();

File: src/python/datastructures/queue/2.queue_with_linkedlist/queue_with_linkedlist.py
"""
Queue implementation using linked list in Python.
"""


class Node:
    """Node class for the queue."""
    
    def __init__(self, value):
        self.value = value
        self.next = None


class MyQueue:
    """Queue implementation using a linked list."""
    
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    # Core queue operations
    def enqueue(self, value):
        """Adds an item to the back of the queue."""
        new_node = Node(value)
        if self.is_empty():
            self.first = new_node
            self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node
        self.length += 1

    def dequeue(self):
        """Removes and returns the front item from the queue."""
        if self.is_empty():
            raise Exception("Queue Underflow: Cannot dequeue from an empty queue.")
        
        temp = self.first
        if self.length == 1:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
            temp.next = None
        
        self.length -= 1
        return temp.value

    def peek(self):
        """Returns the front item without removing it."""
        if self.is_empty():
            raise Exception("Queue Underflow: Cannot peek from an empty queue.")
        return self.first.value

    # Helper functions
    def is_empty(self):
        """Returns True if the queue is empty."""
        return self.length == 0

    def size(self):
        """Returns the number of elements in the queue."""
        return self.length

    def print_queue(self):
        """Returns a string representation of the queue."""
        current = self.first
        values = []
        while current:
            values.append(str(current.value))
            current = current.next
        return ' '.join(values)


# Usage example
if __name__ == "__main__":
    queue = MyQueue()

    print(f"Is empty? {queue.is_empty()}")  # True

    try:
        queue.dequeue()  # Should throw
    except Exception as e:
        print(e)

    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)

    print(f"Queue contents: {queue.print_queue()}")  # 1 2 3
    print(f"Front element: {queue.peek()}")  # 1
    print(f"Dequeued element: {queue.dequeue()}")  # 1
    print(f"Queue after dequeue: {queue.print_queue()}")  # 2 3
